components/api_request_helper/api_request_helper.cc

```diff
@@ -335,106 +315,189 @@ void APIRequestHelper::URLLoaderHandler::send_sse_data_for_testing(
+    VLOG(1) << "Response validation error: Encoding error";
+    std::move(result_callback_).Run(std::move(result));
+    return;
+  }
+  VLOG(2) << "Reponse validation successful";
+  result.body_ = std::move(safe_json);
+  result.value_body_ = std::move(result_value.value());
+  std::move(result_callback_).Run(std::move(result));
+}
+
+void APIRequestHelper::URLLoaderHandler::MaybeSendResult() {
+  // Verify that counting hasn't gone wrong - it should never be less than 0.
+  DCHECK_LT(-1, current_decoding_operation_count_);
+  // Don't allow completion if decoding is still in progress so that
+  // we don't delete the reference to |data_decoder_| which would cancel the
+  // operations.
+  // Response must be complete to know that decoding is definitely complete
+  // since streaming responses may get more reponse chunks which need decoding.
+  const bool decoding_is_complete = (current_decoding_operation_count_ == 0);
+
+  if (request_is_finished_ && decoding_is_complete) {
+    std::move(result_callback_).Run(ToAPIRequestResult(std::move(url_loader_)));
+  } else if (decoding_is_complete) {
+    VLOG(3) << "Did not run URLLoaderHandler completion handler, still have "
+            << current_decoding_operation_count_
+            << " decoding operations in progress, waiting for them to"
+            << " complete...";
+  }
 }
 
 void APIRequestHelper::URLLoaderHandler::ParseSSE(
     base::StringPiece string_piece) {
+  // New chunks should only be received before the request is completed
+  DCHECK(!request_is_finished_);
   // We split the string into multiple chunks because there are cases where
   // multiple chunks are received in a single call.
-  std::vector<std::string> stream_data = base::SplitString(
+  std::vector<base::StringPiece> stream_data = base::SplitStringPiece(
       string_piece, "\r\n", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
 
-  // TODO(@nullhook): Identify both JSON and string values. The below currently
+  // Remove SSE events that don't look like JSON - could be string or [DONE]
+  // message.
+  // TODO(@nullhook): Parse both JSON and string values. The below currently
   // only identifies JSON values.
   static constexpr char kDataPrefix[] = "data: {";
+  stream_data.erase(base::ranges::remove_if(
+                        stream_data,
+                        [](base::StringPiece item) {
+                          DVLOG(3) << "Received chunk: " << item;
+                          if (!base::StartsWith(item, kDataPrefix)) {
+                            // This is useful to log in case an API starts
+                            // coming back with unknown data type in some
+                            // scenarios.
+                            VLOG(1) << "Data did not start with SSE prefix";
+                            return true;
+                          }
+                          return false;
+                        }),
+                    stream_data.end());
```

Comment:
nit: base::EraseIf() combines  erase + remove_if

Reply:
thanks, forgot about that!